backend/fms_core/resources/project_link_sample.py

```diff
@@ -0,0 +1,131 @@
+from import_export.widgets import ForeignKeyWidget
+from django.core.exceptions import ValidationError, ObjectDoesNotExist
+from ._generic import GenericResource
+from ._utils import skip_rows, add_columns_to_preview
+from ..models import Project, Sample, Container, SampleByProject
+from ..utils import get_normalized_str
+
+
+__all__ = ["ProjectLinkSampleResource"]
+
+ADD_ACTION = "ADD"
+REMOVE_ACTION = "REMOVE"
+
+class ProjectLinkSampleResource(GenericResource):
+    action = Field(column_name='Action')
+    project_name = Field(column_name='Project Name') 
+    sample_name = Field(column_name='Sample Name')
+    barcode = Field(column_name='Sample Container Barcode',
+                     widget=ForeignKeyWidget(Container, 'barcode'))
+    coordinates = Field(column_name='Sample Container Coord')
+
+    class Meta:
+        model = SampleByProject
+        fields = (
+            "id",
+        )
+
+
+    def before_import(self, dataset, using_transactions, dry_run, **kwargs):
+        skip_rows(dataset, 7)
+
+    def import_obj(self, obj, data, dry_run, **kwargs):
+        errors = {}
+        # Get field info
+        try:
+            super().import_obj(obj, data, dry_run, **kwargs)
+        except ValidationError as e:
+            errors.update(e.update_error_dict(errors).copy())
+        # Get sample object
+        try:
+            obj.sample = Sample.objects.get(name=data["Sample Name"],
+                                            container__barcode=data["Sample Container Barcode"],
+                                            coordinates=data["Sample Container Coord"])
+        except Sample.DoesNotExist:
+            obj.sample = None
+            errors["sample"] = ValidationError([f"No sample with name [{data['Sample Name']}] at those container coordinates."], code="invalid")
+        # Get project object
+        try:
+            obj.project = Project.objects.get(name=data["Project Name"])
+        except Project.DoesNotExist:
+            obj.project = None
+            errors["project"] = ValidationError([f"No project with name [{data['Project Name']}]"], code="invalid")
+
+        if obj.sample_id and obj.project_id:
+            # Check if link exists to ensure there's not a duplicated association
+            if SampleByProject.objects.filter(sample=obj.sample, project=obj.project).exists() and data["Action"] == ADD_ACTION:
+                errors["link"] = ValidationError(
+                    [f"Sample [{data['Sample Name']}] is already associated to project [{data['Project Name']}]."],
+                    code="invalid")
+
+            # If the link doesn't exists we can't perform a remove action
+            else:
+                if data["Action"] == REMOVE_ACTION:
+                    errors["link"] = ValidationError([f"Sample [{data['Sample Name']}] is not currently associated to project [{data['Project Name']}]."],
+                                                     code="invalid")
```

Comment:
I found this conditional statement rather difficult to read imo. 

In the case where the association already exists, but the action to be performed is REMOVE_ACTION, the if should evaluate to False (True and False => False), we then fall in the else: and there data['Action'] is not REMOVE, so nothing happens but we have to read carefully each conditional branch.

Reply:
I suggest maybe changing the else to an elif data["Action"] == REMOVE_ACTION?

Reply:
Yes, something like: 
if data action == ADD and SampleByProject filter blablab exists   
     do .... 
elif data action == REMOVE 
   do ...